File: tests/05-test_execute_with_session_with_counter_ix.ts
import * as anchor from "@anchor-lang/core";
import { Program } from "@anchor-lang/core";
import { Keypair, PublicKey, TransactionInstruction } from "@solana/web3.js";
import assert from "assert";

import { R3SessionKeys } from "../target/types/r3_session_keys";
import * as sessionKeys from "./utils/pda";
import {
  ensureProgramInitialized,
  executeWithSession,
  futureExpiresAt,
  getOnChainUnixTimestamp,
  loadProgramFromFile,
  sendExpectError,
  setAccountData,
} from "./utils/helpers";
import { ANCHOR_DISCRIMINATOR_LEN, TARGET_PROGRAM_PLACEHOLDER } from "./utils/constants";
import { MockProgram, mockProgramIdl } from "./fixtures/mock-program";
import * as mock from "./fixtures/mock-program";

describe("05 - Execute with session (counter instruction)", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const baseWallet = provider.wallet as anchor.Wallet;
  const connection = provider.connection;

  const program = anchor.workspace.r3SessionKeys as Program<R3SessionKeys>;
  let mockProgram: Program<MockProgram>;

  let mockCounterPda: PublicKey;

  before(async () => {
    await ensureProgramInitialized(program, baseWallet);

    await loadProgramFromFile(connection, mock.MOCK_PROGRAM_ID, mock.MOCK_PROGRAM_SO_PATH);
    mockProgram = new Program(mockProgramIdl as MockProgram, provider) as Program<MockProgram>;
    mockCounterPda = mock.deriveCounterPda(mockProgram.programId);
  });

  /**
   * The mock initialize instruction makes its payer the counter authority.
   * The counter data is overwritten with a cheatcode so the test can exercise PDA signing.
   */
  async function initializeMockCounter(authority: PublicKey) {
    const counterAccount = await connection.getAccountInfo(mockCounterPda);
    if (!counterAccount) {
      await mockProgram.methods
        .initialize()
        .accounts({
          payer: baseWallet.publicKey,
        })
        .signers([baseWallet.payer])
        .rpc()
        .catch((e) => {
          console.log("Error: ", e);
          throw e;
        });
    }

    await setAccountData(connection, mockCounterPda, mock.encodeCounterAccount(0, authority));
  }

  async function createSmartWallet(smartWalletOwner: PublicKey): Promise<PublicKey> {
    const userSmartWalletPda = sessionKeys.deriveUserSmartWalletPda(
      program.programId,
      smartWalletOwner
    );

    await program.methods
      .createSmartWallet(smartWalletOwner)
      .accounts({
        admin: baseWallet.publicKey,
      })
      .signers([baseWallet.payer])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });

    return userSmartWalletPda;
  }

  async function createSession(params: {
    userSmartWallet: PublicKey;
    sessionKey: PublicKey;
    targetProgram: PublicKey;
    expiresAt: anchor.BN;
    allowedInstructionsDiscriminators: Buffer;
    discriminatorLen: number;
  }) {
    await program.methods
      .createSession(
        params.sessionKey,
        params.targetProgram,
        params.expiresAt,
        params.allowedInstructionsDiscriminators,
        params.discriminatorLen
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet: params.userSmartWallet,
      })
      .signers([baseWallet.payer])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });
  }

  async function approveSession(smartWalletOwner: Keypair, sessionKey: PublicKey) {
    await program.methods
      .approveSession(sessionKey)
      .accounts({
        smartWalletOwner: smartWalletOwner.publicKey,
      })
      .signers([baseWallet.payer, smartWalletOwner])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });
  }

  async function revokeSession(smartWalletOwner: Keypair, sessionKey: PublicKey) {
    await program.methods
      .revokeSession(sessionKey)
      .accounts({
        smartWalletOwner: smartWalletOwner.publicKey,
      })
      .signers([baseWallet.payer, smartWalletOwner])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });
  }

  function incrementInstruction(authority: PublicKey): Promise<TransactionInstruction> {
    return mockProgram.methods
      .increment()
      .accounts({
        authority,
      })
      .instruction();
  }

  it("Execute mock increment with session", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = await createSmartWallet(smartWalletOwner.publicKey);

    await initializeMockCounter(userSmartWallet);
    const incrementIx = await incrementInstruction(userSmartWallet);
    assert.ok(incrementIx.data.subarray(0, 8).equals(mock.INCREMENT_DISCRIMINATOR));

    await createSession({
      userSmartWallet,
      sessionKey: sessionKey.publicKey,
      targetProgram: mockProgram.programId,
      expiresAt: await futureExpiresAt(connection),
      allowedInstructionsDiscriminators: mock.INCREMENT_DISCRIMINATOR,
      discriminatorLen: mock.INCREMENT_DISCRIMINATOR.length,
    });
    await approveSession(smartWalletOwner, sessionKey.publicKey);

    await executeWithSession(program, {
      sessionExecutor: baseWallet.publicKey,
      sessionKey: sessionKey.publicKey,
      userSmartWallet,
      targetInstruction: incrementIx,
    })
      .signers([baseWallet.payer, sessionKey])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });

    const counter = await mockProgram.account.counter.fetch(mockCounterPda);
    assert.equal(counter.count.toNumber(), 1);
    assert.equal(counter.authority.toBase58(), userSmartWallet.toBase58());
  });

  it("Execute rejects target program not authorized by session", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = await createSmartWallet(smartWalletOwner.publicKey);

    await initializeMockCounter(userSmartWallet);
    const incrementIx = await incrementInstruction(userSmartWallet);

    // The session is bound to the system program, the executed instruction targets the mock program
    await createSession({
      userSmartWallet,
      sessionKey: sessionKey.publicKey,
      targetProgram: TARGET_PROGRAM_PLACEHOLDER,
      expiresAt: await futureExpiresAt(connection),
      allowedInstructionsDiscriminators: mock.INCREMENT_DISCRIMINATOR,
      discriminatorLen: mock.INCREMENT_DISCRIMINATOR.length,
    });
    await approveSession(smartWalletOwner, sessionKey.publicKey);

    await sendExpectError(
      executeWithSession(program, {
        sessionExecutor: baseWallet.publicKey,
        sessionKey: sessionKey.publicKey,
        userSmartWallet,
        targetInstruction: incrementIx,
      })
        .signers([baseWallet.payer, sessionKey])
        .rpc()
    );
  });

  it("Execute mock increment not allowed by session", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = await createSmartWallet(smartWalletOwner.publicKey);

    await initializeMockCounter(userSmartWallet);
    const incrementIx = await incrementInstruction(userSmartWallet);

    // No discriminator is allowed by this session
    await createSession({
      userSmartWallet,
      sessionKey: sessionKey.publicKey,
      targetProgram: mockProgram.programId,
      expiresAt: await futureExpiresAt(connection),
      allowedInstructionsDiscriminators: Buffer.alloc(0),
      discriminatorLen: ANCHOR_DISCRIMINATOR_LEN,
    });
    await approveSession(smartWalletOwner, sessionKey.publicKey);

    const error = await sendExpectError(
      executeWithSession(program, {
        sessionExecutor: baseWallet.publicKey,
        sessionKey: sessionKey.publicKey,
        userSmartWallet,
        targetInstruction: incrementIx,
      })
        .signers([baseWallet.payer, sessionKey])
        .rpc()
    );

    assert.ok(error.includes("NotAllowedInstructionDiscriminator"), error);
    const counter = await mockProgram.account.counter.fetch(mockCounterPda);
    assert.equal(counter.count.toNumber(), 0);
    assert.equal(counter.authority.toBase58(), userSmartWallet.toBase58());
  });

  it("Execute mock increment with expired session key", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = await createSmartWallet(smartWalletOwner.publicKey);

    await initializeMockCounter(userSmartWallet);
    const incrementIx = await incrementInstruction(userSmartWallet);

    // Expiration is exclusive, so a session expiring at the current timestamp is expired.
    const expiresAt = new anchor.BN(await getOnChainUnixTimestamp(connection));
    await createSession({
      userSmartWallet,
      sessionKey: sessionKey.publicKey,
      targetProgram: mockProgram.programId,
      expiresAt,
      allowedInstructionsDiscriminators: mock.INCREMENT_DISCRIMINATOR,
      discriminatorLen: mock.INCREMENT_DISCRIMINATOR.length,
    });
    await approveSession(smartWalletOwner, sessionKey.publicKey);

    const error = await sendExpectError(
      executeWithSession(program, {
        sessionExecutor: baseWallet.publicKey,
        sessionKey: sessionKey.publicKey,
        userSmartWallet,
        targetInstruction: incrementIx,
      })
        .signers([baseWallet.payer, sessionKey])
        .rpc()
    );

    assert.ok(error.includes("SessionExpired"), error);
    const counter = await mockProgram.account.counter.fetch(mockCounterPda);
    assert.equal(counter.count.toNumber(), 0);
    assert.equal(counter.authority.toBase58(), userSmartWallet.toBase58());
  });

  it("Execute mock increment with revoked session", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = await createSmartWallet(smartWalletOwner.publicKey);

    await initializeMockCounter(userSmartWallet);
    const incrementIx = await incrementInstruction(userSmartWallet);

    await createSession({
      userSmartWallet,
      sessionKey: sessionKey.publicKey,
      targetProgram: mockProgram.programId,
      expiresAt: await futureExpiresAt(connection),
      allowedInstructionsDiscriminators: mock.INCREMENT_DISCRIMINATOR,
      discriminatorLen: mock.INCREMENT_DISCRIMINATOR.length,
    });
    await approveSession(smartWalletOwner, sessionKey.publicKey);
    await revokeSession(smartWalletOwner, sessionKey.publicKey);

    const error = await sendExpectError(
      executeWithSession(program, {
        sessionExecutor: baseWallet.publicKey,
        sessionKey: sessionKey.publicKey,
        userSmartWallet,
        targetInstruction: incrementIx,
      })
        .signers([baseWallet.payer, sessionKey])
        .rpc()
    );

    assert.ok(error.includes("SessionNotApproved"), error);
    const counter = await mockProgram.account.counter.fetch(mockCounterPda);
    assert.equal(counter.count.toNumber(), 0);
    assert.equal(counter.authority.toBase58(), userSmartWallet.toBase58());
  });
});

File: programs/r3-session-keys/tests/05-test_execute_with_session_with_counter_ix.rs
mod common;

use {
    anchor_lang::{prelude::Clock, AccountSerialize},
    common::{
        client, load, mock_client, send_tx_expect_error, send_tx_expect_ok, Env,
        TARGET_PROGRAM_PLACEHOLDER,
    },
    mock_client::mock_program::accounts::Counter,
    solana_keypair::Keypair,
    solana_signer::Signer,
    std::time::{SystemTime, UNIX_EPOCH},
};

fn load_mock_program(env: &mut Env) {
    let bytes = include_bytes!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../tests/fixtures/mock-program/mock_program.so"
    ));
    env.svm
        .add_program(mock_client::mock_program::ID, bytes)
        .unwrap();
}

fn initialize_mock_counter(
    env: &mut Env,
    mock_client: &mock_client::MockClient,
    authority: anchor_lang::prelude::Pubkey,
) {
    let counter = mock_client::counter_pda();
    let initialize_ix = mock_client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    // The mock initialize instruction makes its payer the counter authority.
    // LiteSVM changes only that field so the test can exercise PDA signing.
    let mut counter_account = env.svm.get_account(&counter).unwrap();
    let counter_state = Counter {
        count: 0,
        authority,
    };
    let mut account_data = counter_account.data.as_mut_slice();
    counter_state.try_serialize(&mut account_data).unwrap();
    env.svm.set_account(counter, counter_account).unwrap();
}

#[test]
fn test_execute_mock_increment_with_session() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    load_mock_program(&mut env);
    let mock_client = mock_client::MockClient::new();

    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);

    initialize_mock_counter(&mut env, &mock_client, user_smart_wallet);
    let mock_counter = mock_client::counter_pda();
    let increment_ix = mock_client.increment(user_smart_wallet);
    let increment_discriminator = mock_client::increment_discriminator().to_vec();
    assert!(increment_ix.data.starts_with(&increment_discriminator));

    let expires_at = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64
        + 1000;
    let (create_session, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        mock_client::mock_program::ID,
        expires_at,
        increment_discriminator.clone(),
        u8::try_from(increment_discriminator.len()).unwrap(),
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let approve_session = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let execute = client.execute_with_session(
        env.admin.pubkey(),
        session_key.pubkey(),
        user_smart_wallet,
        increment_ix,
    );
    send_tx_expect_ok(&mut env.svm, execute, &[&env.admin, &session_key]);

    let counter: Counter = load(&env, &mock_counter);
    assert_eq!(counter.count, 1);
    assert_eq!(counter.authority, user_smart_wallet);
}

#[test]
fn test_execute_rejects_target_program_not_authorized_by_session() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    load_mock_program(&mut env);
    let mock_client = mock_client::MockClient::new();

    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);

    initialize_mock_counter(&mut env, &mock_client, user_smart_wallet);
    let increment_ix = mock_client.increment(user_smart_wallet);
    let increment_discriminator = mock_client::increment_discriminator().to_vec();

    let expires_at = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64
        + 1000;
    let (create_session, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        TARGET_PROGRAM_PLACEHOLDER,
        expires_at,
        increment_discriminator.clone(),
        u8::try_from(increment_discriminator.len()).unwrap(),
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let approve_session = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let execute = client.execute_with_session(
        env.admin.pubkey(),
        session_key.pubkey(),
        user_smart_wallet,
        increment_ix,
    );
    send_tx_expect_error(&mut env.svm, execute, &[&env.admin, &session_key]);
}

#[test]
fn test_execute_mock_increment_not_allowed_by_session() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    load_mock_program(&mut env);
    let mock_client = mock_client::MockClient::new();

    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);

    initialize_mock_counter(&mut env, &mock_client, user_smart_wallet);
    let mock_counter = mock_client::counter_pda();
    let increment_ix = mock_client.increment(user_smart_wallet);

    let expires_at = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64
        + 1000;
    let (create_session, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        mock_client::mock_program::ID,
        expires_at,
        vec![],
        u8::try_from(mock_client::increment_discriminator().len()).unwrap(),
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let approve_session = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let execute = client.execute_with_session(
        env.admin.pubkey(),
        session_key.pubkey(),
        user_smart_wallet,
        increment_ix,
    );
    let error = send_tx_expect_error(&mut env.svm, execute, &[&env.admin, &session_key]);

    assert!(
        error.contains("Error Code: NotAllowedInstructionDiscriminator"),
        "{error}"
    );
    let counter: Counter = load(&env, &mock_counter);
    assert_eq!(counter.count, 0);
    assert_eq!(counter.authority, user_smart_wallet);
}

#[test]
fn test_execute_mock_increment_with_expired_session_key() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    load_mock_program(&mut env);
    let mock_client = mock_client::MockClient::new();

    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);

    initialize_mock_counter(&mut env, &mock_client, user_smart_wallet);
    let mock_counter = mock_client::counter_pda();
    let increment_ix = mock_client.increment(user_smart_wallet);
    let increment_discriminator = mock_client::increment_discriminator().to_vec();

    // Expiration is exclusive, so a session expiring at the current timestamp is expired.
    let expires_at = env.svm.get_sysvar::<Clock>().unix_timestamp;
    let (create_session, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        mock_client::mock_program::ID,
        expires_at,
        increment_discriminator.clone(),
        u8::try_from(increment_discriminator.len()).unwrap(),
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let approve_session = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let execute = client.execute_with_session(
        env.admin.pubkey(),
        session_key.pubkey(),
        user_smart_wallet,
        increment_ix,
    );
    let error = send_tx_expect_error(&mut env.svm, execute, &[&env.admin, &session_key]);

    assert!(error.contains("Error Code: SessionExpired"), "{error}");
    let counter: Counter = load(&env, &mock_counter);
    assert_eq!(counter.count, 0);
    assert_eq!(counter.authority, user_smart_wallet);
}

#[test]
fn test_execute_mock_increment_with_revoked_session() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    load_mock_program(&mut env);
    let mock_client = mock_client::MockClient::new();

    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);

    initialize_mock_counter(&mut env, &mock_client, user_smart_wallet);
    let mock_counter = mock_client::counter_pda();
    let increment_ix = mock_client.increment(user_smart_wallet);
    let increment_discriminator = mock_client::increment_discriminator().to_vec();

    let expires_at = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64
        + 1000;
    let (create_session, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        mock_client::mock_program::ID,
        expires_at,
        increment_discriminator.clone(),
        u8::try_from(increment_discriminator.len()).unwrap(),
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let approve_session = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let revoke_session = client.revoke_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        revoke_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let execute = client.execute_with_session(
        env.admin.pubkey(),
        session_key.pubkey(),
        user_smart_wallet,
        increment_ix,
    );
    let error = send_tx_expect_error(&mut env.svm, execute, &[&env.admin, &session_key]);

    assert!(error.contains("Error Code: SessionNotApproved"), "{error}");
    let counter: Counter = load(&env, &mock_counter);
    assert_eq!(counter.count, 0);
    assert_eq!(counter.authority, user_smart_wallet);
}
